### PR TITLE
Multiple improvements for programming exercises

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
`Trigger all builds` and `build and test after due date` did not work for all participations in large courses due to problems with `resume exercise` and with transactions. We carefully inspected the problems and implemented many small improvements so that these features work as expected

### Description
* Remove most `Transactional` annotations for programming exercises. We don't really need transactions here. This makes the execution faster and less error-prone in terms of concurrency issues
* Fix concurrency issues during `Trigger all builds` and `build and test after due date`
* Change the order of empty commits for resume exercise
* Make sure students can always resume programming exercises, even if a step fails
* Improve error handling
* Filter all branches except master and Artemis setup commits to prevent wrong submissions
* Use the student / instructor user name and email for commits in the online code editor and when generating the exercise
* Handle go to build plan on client to prevent null pointer exceptions

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise
4. Start exercise in Courses
5. Clean up the build plan
6. Run Trigger All in Participations
7. Check that all builds have been executed

**Note:** All other aspects of programming exercises need to be tested too.
We need to simulate the trigger all button with 1000 test users on the test server

### CODE REVIEW

File: src/main/webapp/app/entities/result/updating-result.component.ts [hunk 1]
```diff
@@ -73,6 +73,16 @@ export class UpdatingResultComponent implements OnChanges, OnDestroy {
             .pipe(
                 // Ignore initial null result of subscription
                 filter(result => !!result),
+                // TODO: remove logging
```
Comment (jpbernius) [inline src/main/webapp/app/entities/result/updating-result.component.ts]: - [ ] remove logging

File: src/main/webapp/app/scores/exercise-scores.component.ts [hunk 2]
```diff
@@ -207,7 +207,8 @@ export class ExerciseScoresComponent implements OnInit, OnDestroy {
     }
 
     goToBuildPlan(result: Result) {
-        this.sourceTreeService.goToBuildPlan(result.participation!);
+        // TODO: get the continuous integration URL as a client constant during the management info call
+        window.open('https://bamboobruegge.in.tum.de/browse/' + (result.participation! as ProgrammingExerciseStudentParticipation).buildPlanId);
```
Comment (jpbernius) [inline src/main/webapp/app/scores/exercise-scores.component.ts]: Is this the way to go? Especially if we are looking at an upcoming Jenkins integration, I think it is easier to build up this URL server-side.

**Comment by jpbernius** [comment]
For me in Exercise [atest24](https://artemistest.ase.in.tum.de/#/overview/1/exercises/1026), I can only submit once in the code editor, the button becomes inactive. I can save, but not submit. After reloading the page, I can submit again.

File: src/main/webapp/app/entities/result/updating-result.component.ts (hunk 1 above)
Comment (krusche) [reply src/main/webapp/app/entities/result/updating-result.component.ts]: the logging is here to investigate why the submit button in the code editor does not work for the second time. I guess something is missing when we send the result to the client. Did you notice something in the console logs?

File: src/main/webapp/app/scores/exercise-scores.component.ts (hunk 2 above)
Comment (krusche) [reply src/main/webapp/app/scores/exercise-scores.component.ts]: We have lots of null pointer exceptions on the server. As we have this hard coded in many places, it does not make sense to have a server call here that might even fail.

I think the TODO is the way to go. Load the BASE URL for the CI Service once and then build the URL on the client without additional server calls. However this is not important for now, we can fix this in a future PR

File: src/main/webapp/app/entities/result/updating-result.component.ts (hunk 1 above)
Comment (jpbernius) [reply src/main/webapp/app/entities/result/updating-result.component.ts]: No, nothing in the browser logs 😔

**Comment by krusche** [comment]
> For me in Exercise [atest24](https://artemistest.ase.in.tum.de/#/overview/1/exercises/1026), I can only submit once in the code editor, the button becomes inactive. I can save, but not submit. After reloading the page, I can submit again.

I fixed the problem. Please test again!

**Comment by thilo-behnke** [comment]
Very nice! I think this will enable a more stable conduction of the programming exercises in future.
I don't know how / if we should look into issue 2 as it seems to work in general, but I'll look into issues 1 and 3 now.

My findings:

1) Sequential programming exercises are broken as feedbacks can't be saved (order_column issue, probably because we removed the transactional annotation):

E.g. #/code-editor/1119/admin/155328

https://sentry.io/organizations/ls1intum/issues/1318627036/?environment=test&project=1440029&query=is%3Aunresolved

2) After commit in code-editor no submission is sent to the client (Bug reported by Jan-Philipp)

Only on test server for some exercises, e.g. #/code-editor/155321.
I don't understand what the reproduction criteria are, there are no errors in Sentry.

3) Trigger all builds generates a failed submission for the template participation. 

Trying to trigger the failed build again for the template causes a 400 as the last submission was not failed. Seems like a client side issue, the submission is successful when reloading the page.

**Comment by thilo-behnke** [comment]
FYI: The sequential test runs did not work anymore, because of the way the Result with the Feedback items is saved. We don't save the Result in BambooService anymore, but only when it is updated with the TestCase information (weights, etc.).

**Pull request merged**